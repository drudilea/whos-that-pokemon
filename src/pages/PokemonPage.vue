<template>
  <div v-if="!pokemon">
    <h1>Espere por favor...</h1>
  </div>

  <div v-else>
    <h1>¿Quién es este pokémon?</h1>

    <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon" />
    <PokemonOptions
      :pokemons="pokemonsArr"
      @selectedPokemon="checkAnswer($event)"
    />

    <template v-if="showAnswer">
      <h2>{{ message }}</h2>
      <button @click="newGame">Nuevo juego</button>
    </template>
  </div>
</template>

<script>
import PokemonOptions from '@/components/PokemonOptions';
import PokemonPicture from '@/components/PokemonPicture';

import getPokemonOptions from '@/helpers/getPokemonOptions';

export default {
  components: { PokemonOptions, PokemonPicture },
  data() {
    return {
      pokemonsArr: [],
      pokemon: null,
      showPokemon: false,
      showAnswer: false,
      message: '',
    };
  },
  methods: {
    async mixPokemonsArray() {
      this.pokemonsArr = await getPokemonOptions();
      const rndInt = Math.floor(Math.random() * 4);

      this.pokemon = this.pokemonsArr[rndInt];
    },
    checkAnswer(selectedId) {
      this.showPokemon = true;
      this.showAnswer = true;

      if (selectedId === this.pokemon.id) {
        this.message = `Correcto! Es ${this.pokemon.name}`;
      } else {
        this.message = `Oops! Era ${this.pokemon.name}`;
      }
    },
    newGame() {
      this.initVariables();
      this.mixPokemonsArray();
    },
    initVariables() {
      this.pokemonsArr = [];
      this.pokemon = null;
      this.showPokemon = false;
      this.showAnswer = false;
      this.message = '';
    },
  },
  mounted() {
    this.mixPokemonsArray();
  },
};
</script>
