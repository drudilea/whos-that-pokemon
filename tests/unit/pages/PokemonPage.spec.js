import { shallowMount, mount } from '@vue/test-utils';
import PokemonPage from '@/pages/PokemonPage';
import { pokemons } from '../mocks/pokemons.mock';

describe('PokemonPage Component', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(PokemonPage);
  });

  test('should match the snapshot when pokemons are already loaded', () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonsArr: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: '',
        };
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  test('should call mixPokemonsArray when mounted', () => {
    const mixPokemonsArray = jest.spyOn(
      PokemonPage.methods,
      'mixPokemonsArray'
    );
    shallowMount(PokemonPage);

    expect(mixPokemonsArray).toHaveBeenCalled();
  });

  test('should render both components, PokemonPicture and PokemonOptions', () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonsArr: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: '',
        };
      },
    });

    const pokemonPicture = wrapper.find('pokemon-picture-stub');
    const pokemonOptions = wrapper.find('pokemon-options-stub');

    expect(pokemonPicture.exists()).toBeTruthy();
    expect(pokemonOptions.exists()).toBeTruthy();

    expect(pokemonPicture.attributes('pokemonid')).toBe(
      pokemons[0].id.toString()
    );
    expect(pokemonOptions.attributes('pokemons')).toBeTruthy();
  });

  test('checkAnswer tests', async () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonsArr: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: '',
        };
      },
    });

    await wrapper.vm.checkAnswer(pokemons[0].id);

    expect(wrapper.find('h2').exists()).toBeTruthy();
    expect(wrapper.find('h2').text()).toBe(`Correcto! Es ${pokemons[0].name}`);
    expect(wrapper.vm.showPokemon).toBe(true);

    await wrapper.vm.checkAnswer(pokemons[2].id);
    expect(wrapper.vm.message).toBe(`Oops! Era ${pokemons[0].name}`);
  });
});
