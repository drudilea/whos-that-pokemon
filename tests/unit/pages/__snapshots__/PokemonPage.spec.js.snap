// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Component should match the snapshot when pokemons are already loaded 1`] = `
<div>
  <h1>¿Quién es este pokémon?</h1>
  <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
